edb/schema/delta.py

```diff
@@ -2176,12 +2176,50 @@ def _get_expr_field_ast(
         parent_cls = parent_op.get_schema_metaclass()
         has_shadow = parent_cls.get_field(f'orig_{field.name}') is not None
 
-        if context.descriptive_mode:
-            # When generating AST for DESCRIBE AS TEXT, we want
-            # to use the original user-specified and unmangled
-            # expression to render the object definition.
-            expr_ql = edgeql.parse_fragment(self.new_value.origtext)
-        else:
+        expr_ql = None
+        if context.descriptive_mode or context.declarative:
+            # Potentially we want to use the original expression
+            # instead of the normalized one, but this can only be done
+            # safely if the normalized form has sets the same module
+            # as the one that hosts the object we're in.
+            #
+            # We also make a special exception for indexes which
+            # should always have a valid original expression since
+            # they cannot use functions.
+            #
+            # When generating AST for DESCRIBE AS TEXT or AS SDL, we
+            # want to use the original user-specified and unmangled
+            # expression to render the object definition whenever
+            # possible.
+
+            # # index is going to be that exact class, not even a subclass
+            # from . import indexes as s_indexes
+            # if parent_cls is s_indexes.Index:
+            #     expr_ql = edgeql.parse_fragment(self.new_value.origtext)
+
+            if isinstance(self.new_value.qlast, qlast.SelectQuery):
+                aliases = self.new_value.qlast.aliases
+                # now need to make sure this looks like normalized form
+                num_aliases = len(aliases)
+
+                if (num_aliases == 0 or
+                    # No aliases at all or a single module alias are
+                    # both indicators that we might be able to use
+                    # orig_expr.
+                    (num_aliases == 1
+                        and isinstance(aliases[0], qlast.ModuleAliasDecl)
+                        # and that the module matches
+                        and aliases[0].module == parent_op.classname.module)):
+
+                    expr_ql = edgeql.parse_fragment(self.new_value.origtext)
```

Comment:
I'm not sure we need this stuff. Too much guessing and I don't like that.

@elprans your thoughts?

Reply:
In defence of this, the main point is that we can rely on the normalization code to create a fairly predictable query here and if the query doesn't fit this, it pretty much implies that it wasn't normalized so it doesn't need adjustments.

Reply:
This should also check if `origtext` != `text` to account for cases when `WITH MODULE` was actually specified by the user.  I would also factor this check out and place it near `imprint_expr_context()` to make sure they don't go out of sync.

Reply:
> @elprans your thoughts?

I think that by adding `and origtext != text` to the condition the check is made unambiguous.  Implementing it as a helper near `imprint_source_context()`  rather than open-coding here should further solidify this.